pypeclub/OpenPype
# Workfiles: Open published workfiles

## Brief description
It is handy to have ability to open published workfile through Workfiles tool in some cases.

## Description
We've agreed that we don't want to allow direct access to published workfile but want add ability to open it. To avoid opening of published workfile directly we decided to create a copy of workfile on user's machine and then open it. This won't close workfiles tool so it is possible to save it right away to a selected context.

Because this way there are created files in app data there should be ability to remove them without user's direct interaction. That is achieved using simple metadata cache mechanism that will store each opened workfile into appdata and store metadata with it's expiration time (1 hour right now). Cache is validated on initialization of workfiles tool which may not be enough so it is also possible to clear cache manually using button in workfiles tool. Next to clear temp is information about used disk storage.

Button label in "publish mode" has label `View` (any better suggestions?). When any published workfile is opened it is copied to cache folder, opened from there using host implementation and `Published` checkbox is unchecked.

### How are selected workfiles?
All subsets with all versions and all representations are loaded for selected asset (it may take some time to load). First filtering is by extensions in representation context which must match one of workfile extensions of host. When task is selected it is also checked for task name in the context. All of them are added to view.

### Screenshot
![image](https://user-images.githubusercontent.com/43494761/159310402-65ccccf8-e844-4378-8f69-f5aa162dee40.png)

## Changes
- workfiles tool uses `QStandardItemModel` for filling data
- widgets in workfiles tool were separated into more files (it was hard to work in single file)
- widget `FilesWidget` has 2 views and models
    - one is for workarea files which works same way as did before
    - second is for published files which loads paths of representations matching current context

## Testing notes:
1. Open workfiles tool in any host
2. Check `Published` checkbox
3. Open any published workfile
4. `Published` checkbox should be unchecked
5. Now you can store opened file into any context
6. The cached file in appsdata should be removed over time on it's own

Modified implementation of PR https://github.com/pypeclub/OpenPype/pull/2442/files
